### PR TITLE
Create the registration page

### PR DESCRIPTION
**Pull request by casperjureskog**
**PT Story:** https://www.pivotaltracker.com/story/show/141602857

**Changes proposed in this PR:**

* Adds devise user registration page
* Adds rspec and cucumber tests

**What I have learned working on this feature:**
That devise have some very special helpers

**Screenshots:**
screen shot 2017-03-14 at 22 31 56

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
```
Comment (Scharrre97) [inline app/models/user.rb]: please remove comments.

File: config/environments/development.rb [hunk 2]
```diff
@@ -31,6 +31,8 @@
 
   config.action_mailer.perform_caching = false
 
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+
   # Print deprecation notices to the Rails logger.
```
Comment (Scharrre97) [inline config/environments/development.rb]: please remove comments.

File: features/user_registration.feature [hunk 3]
```diff
@@ -0,0 +1,14 @@
+Feature: As a visitor
+In order to checkout
+I would like to create an account
+
+
+Scenario: I create an account
```
Comment (sigu) [inline features/user_registration.feature]: When will the sad path for this be done?

Comment (casperjureskog) [reply features/user_registration.feature]: not at this moment we are focusing on geting the site upp and running :-)

Comment (AmberWilkie) [reply features/user_registration.feature]: Love it, @sigu. Yes! @casperjureskog this should definitely have a sad path. Try making a new scenario when the user doesn't fill in email. You'll be able to use the built-in devise flash messages to do like `I should see "Email is required"`

**Comment by AmberWilkie** [comment]
Unfortunately some changes from the Images PR messed with our tests here. @casperjureskog this will need a second look and to fix merge conflicts. I would suggest we just move our "Welcome whatever" to the `images` page, which is the new root. The other conflicts should be ok to deal with.

File: features/user_registration.feature (hunk 3 above)
Comment (tochman) [reply features/user_registration.feature]: Okay, so @casperjureskog should probably articulate his thoughts a little better but what we discussed is that we should try to focus on happy paths for the MVP and disregard the corner cases (at least when we need to implement those solutions).  I think that this is generally a good idea. In this case though, since Devise is giving us a lot of functionality out-of--the-box, we should probably add tests for sad paths.

**Comment by tochman** [comment]
After reviewing this I would like you to remove the views we are not using in this application. Overall, I would suggest that we only keep the files we are actually making use of - I created a ticket for other files I feel needs to be removed.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,3 +1,10 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
```
Comment (tochman) [inline app/controllers/application_controller.rb:8]: Why are we adding name? I can't find anything about this attribute in the Tracker or on any lo fi's

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+
+
```
Comment (tochman) [inline app/models/user.rb]: remove empty lines.. One is enough.

File: app/views/devise/confirmations/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email) %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Resend confirmation instructions" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (tochman) [inline app/views/devise/confirmations/new.html.erb]: Not used...

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
+
+<p>You can confirm your account email through the link below:</p>
+
+<p><%= link_to 'Confirm my account', confirmation_url(@resource, confirmation_token: @token) %></p>
```
Comment (tochman) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: Not used...

File: app/views/devise/mailer/password_change.html.erb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<p>Hello <%= @resource.email %>!</p>
+
+<p>We're contacting you to notify you that your password has been changed.</p>
```
Comment (tochman) [inline app/views/devise/mailer/password_change.html.erb]: Not used...

File: app/views/devise/mailer/reset_password_instructions.html.erb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<p>Hello <%= @resource.email %>!</p>
+
+<p>Someone has requested a link to change your password. You can do this through the link below.</p>
+
+<p><%= link_to 'Change my password', edit_password_url(@resource, reset_password_token: @token) %></p>
+
+<p>If you didn't request this, please ignore this email.</p>
+<p>Your password won't change until you access the link above and create a new one.</p>
```
Comment (tochman) [inline app/views/devise/mailer/reset_password_instructions.html.erb]: Not used...

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -9,6 +9,8 @@
   </head>
 
   <body>
+    <p class="notice"><%= notice %></p>
```
Comment (tochman) [inline app/views/layouts/application.html.erb:12]: Where is the css class `notice` defined?

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -9,6 +9,8 @@
...
+    <p class="alert"><%= alert %></p>
```
Comment (tochman) [inline app/views/layouts/application.html.erb:13]: Same here, where do we define `alert`?

File: spec/factories/users.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :user do
+    email 'email@email.com'
+    name 'nisse'
```
Comment (tochman) [inline spec/factories/users.rb]: I really would prefer if you guys use capital letters in names. `nisse` would probably prefer to be called `Nisse`

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  describe 'validations' do
+    it { is_expected.to validate_presence_of :name }
+    it { is_expected.to validate_presence_of :email }
+  end
+
+  describe 'Factory' do
+    it 'is expected to have a valid factory' do
+      expect(FactoryGirl.create(:user)).to be_valid
```
Comment (tochman) [inline spec/models/user_spec.rb]: I would suggest to use the shorter:
```
 expect(create(:user)).to be_valid
```

**Review by tochman — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (casperjureskog) [reply app/controllers/application_controller.rb:8]: https://www.pivotaltracker.com/story/show/141603563

As a visitor
In order to register my email, **name** and password
I would like to see a registration form with fields

File: spec/models/user_spec.rb (hunk 13 above)
Comment (casperjureskog) [reply spec/models/user_spec.rb]: try it. the test didn't pass :-(

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (tochman) [reply app/controllers/application_controller.rb:8]: Good, thank you. I missed that.

File: spec/models/user_spec.rb (hunk 13 above)
Comment (tochman) [reply spec/models/user_spec.rb]: I would like you to try again...

Comment (tochman) [reply spec/models/user_spec.rb]: I notice that you don't have the latest code merged in.

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (casperjureskog) [reply app/views/layouts/application.html.erb:12]: css added

File: app/views/layouts/application.html.erb (hunk 11 above)
Comment (casperjureskog) [reply app/views/layouts/application.html.erb:13]: css added

File: app/views/devise/passwords/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    <%= f.label :password_confirmation, "Confirm new password" %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "off" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Change my password" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (tochman) [inline app/views/devise/passwords/edit.html.erb]: This view is also not used....

File: app/views/devise/passwords/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Send me reset password instructions" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (tochman) [inline app/views/devise/passwords/new.html.erb]: This view is also not used....

File: app/views/devise/registrations/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  <div class="actions">
+    <%= f.submit "Update" %>
+  </div>
+<% end %>
+
+<h3>Cancel my account</h3>
+
+<p>Unhappy? <%= button_to "Cancel my account", registration_path(resource_name), data: { confirm: "Are you sure?" }, method: :delete %></p>
+
+<%= link_to "Back", :back %>
```
Comment (tochman) [inline app/views/devise/registrations/edit.html.erb]: This view is also not used....

File: app/views/devise/unlocks/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Resend unlock instructions" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (tochman) [inline app/views/devise/unlocks/new.html.erb]: This view is also not used....

**Review by tochman — CHANGES_REQUESTED** [review]
There are still issues with this PR.

File: app/assets/stylesheets/custom.css.scss [hunk 18]
```diff
@@ -14,3 +14,60 @@
...
+a:hover {
+    color: #34DCC7;
+    text-decoration: none;
+}
+
+
+a:active {
+    color: #315954;
+    text-decoration: none;
+}
```
Comment (tochman) [inline app/assets/stylesheets/custom.css.scss]: I am not sure about adding this css in this PR. There exits another story for that in the tracker. Feels out of scope.

File: app/views/layouts/application.html.erb (hunk 11 above)
Comment (casperjureskog) [reply app/views/layouts/application.html.erb:13]: in devise.en.yml and custom.css.scss

Comment (tochman) [reply app/views/layouts/application.html.erb:13]: I was referring to the css class.

**Comment by tochman** [comment]
I have no objections to this PR. IMO it can be merged.

**Comment by diraulo** [comment]
@tochman You'll have to approve this using the review tool

**Review by tochman — APPROVED** [review]

**Review by diraulo — APPROVED** [review]
